### PR TITLE
Improve Btrfs subvols management

### PR DESCRIPTION
**Pull request by joseivanlopez**
## Problem

Btrfs subvolumes are not well integrated in the Expert Partitioner UI. To see the subvolumes of a Btrfs you have to edit the filesystem and use the *Handle Subvolumes* button. It is not easy to simple check the list of subvolumes without editing the filesystem. Moreover, the dialog to create/edit a subvolume is open as a popup, instead of a wizard like in the rest of devices.

current-subvols

This PR is about improving the integration of the Btrfs subvolumes in the Expert Partitioner. The idea is to convert Btrfs subvolumes to first class citizen, similar to partitions or LVM LVs.

* PBI: https://trello.com/c/OQGxR8Eo/2077-5-partitioner-more-integrated-subvolumes-management.

Related bugs:

* https://bugzilla.suse.com/show_bug.cgi?id=996598
* https://bugzilla.suse.com/show_bug.cgi?id=1175139
* https://bugzilla.suse.com/show_bug.cgi?id=1161408

## Solution

Btrfs subvolumes are now presented in a similar way to other devices like partitions or LVM logical volumes:

* In the tables, their are nested to its filesystem (or to the block device when the filesystem is single-device).
* There is a new single-step wizard to create/edit a subvolume.
* Buttons and menus are adapted to also work with a Btrfs subvolume.
* A description for the subvolume is shown when using "Device->Show Details" menu.

Because sometimes there could be too many subvolumes (e.g., when snapper is used), a table entry is automatically closed by default when it has more than 10 children. This affects to all entries, for example a disk with many partitions, no only to subvolumes. 

Bonus: some refactoring was done to buttons and actions.

## Testing

- Added new unit tests.
- Tested manually with the *partitioner_testing* client.

## Screenshots

<details><summary>Show/Hide</summary>

Screenshot from 2020-11-05 12-37-34

Screenshot from 2020-11-05 12-38-34

Screenshot from 2020-11-05 12-39-08

Screenshot from 2020-11-05 12-39-44

Screenshot from 2020-11-05 12-40-43

Screenshot from 2020-11-05 12-44-27
</details>

### CODE REVIEW

File: src/lib/y2storage/filesystems/blk_filesystem.rb [hunk 1]
```diff
@@ -137,6 +137,47 @@ def blk_device_basename
...
 
+      # Display name to represent the filesystem
+      #
+      # Only multidevice filesystems have a display name
+      #
+      # @return [String, nil]
+      def display_name
+        return nil unless multidevice?
+
+        textdomain "storage"
```
Comment (ancorgs) [inline src/lib/y2storage/filesystems/blk_filesystem.rb:148]: Why this call to textdomain here?

**Comment by ancorgs** [comment]
A general note. Now that the branch `partitioner-ui-02` is already merged into master, shouldn't we stop accumulating stuff there (even delete that branch) and create a new one for all these subvolume-related pull requests?

**Comment by joseivanlopez** [comment]
> A general note. Now that the branch `partitioner-ui-02` is already merged into master, shouldn't we stop accumulating stuff there (even delete that branch) and create a new one for all these subvolume-related pull requests?

Done

**Comment by ancorgs** [comment]
I'm not sure if grouping all the buttons that used to be in separate files into the new `device_edit_buttons.rb`, `device_add_buttons.rb` and `device_delete_buttons.rb` files is really a step forward.

File: src/lib/y2partitioner/widgets/configurable_blk_devices_table.rb [hunk 2]
```diff
@@ -152,12 +152,7 @@ def show_columns(*column_names)
       # @macro seeAbstractWidget
       # @see #columns_help
       def help
-        header = _(
-          "<p>This view shows storage devices.</p>" \
-          "<p>The overview contains:</p>" \
-        )
-
-        header + columns_help
+        "<p>Table containing the list of selected devices, where:</p>" + columns_help
```
Comment (ancorgs) [inline src/lib/y2partitioner/widgets/configurable_blk_devices_table.rb:155]: Not sure if "list of selected devices" is really precise and understandable here.

File: src/lib/y2partitioner/widgets/blk_devices_table.rb [hunk 3]
```diff
@@ -55,7 +55,9 @@ def items
       #
       # @return [Hash{String => Boolean}]
       def open_items
-        @open_items || {}
+        return default_open_items if !@open_items || @open_items.empty?
```
Comment (ancorgs) [inline src/lib/y2partitioner/widgets/blk_devices_table.rb:58]: Why is `default_open_items` called when `@open_items == {}`? I understand the nil case, but the empty one looks quite unexpected to me.

File: src/lib/y2partitioner/widgets/description_section/filesystem.rb [hunk 4]
```diff
@@ -142,6 +142,17 @@ def btrfs_metadata_raid_level_value
           format(_("Metadata RAID Level: %s"), level)
         end
 
+        # Information about the devices used by the Btrfs filesystem
+        #
+        # @return [String]
+        def btrfs_devices_value
+          devices = filesystem.plain_blk_devices.map(&:name)
+
+          # TRANSLATORS: Information abut devices used by a Btrfs filesystem, where %s is replaced by a
```
Comment (ancorgs) [inline src/lib/y2partitioner/widgets/description_section/filesystem.rb:151]: typo: abut

**Review by ancorgs — COMMENTED** [review]
Added several small comments.

**Comment by joseivanlopez** [comment]
> I'm not sure if grouping all the buttons that used to be in separate files into the new `device_edit_buttons.rb`, `device_add_buttons.rb` and `device_delete_buttons.rb` files is really a step forward.

They are really small classes. Basically, they only redefine `#label` and `#action`, which are one liner methods. There are some similar cases in `CWM`, for example https://github.com/yast/yast-yast2/blob/master/library/cwm/src/lib/cwm/common_widgets.rb.

In general, I also prefer one file per class, but I am not sure if having all these tiny classes in its own file really worth.

File: src/lib/y2partitioner/widgets/configurable_blk_devices_table.rb (hunk 2 above)
Comment (joseivanlopez) [reply src/lib/y2partitioner/widgets/configurable_blk_devices_table.rb:155]: Yes, I am not sure either. Do you have any idea to improve it?

File: src/lib/y2storage/filesystems/blk_filesystem.rb (hunk 1 above)
Comment (joseivanlopez) [reply src/lib/y2storage/filesystems/blk_filesystem.rb:148]: After discussing it, we have decided to keep it as it is. Maybe in the future we should start using decorators.

**Comment by joseivanlopez** [comment]
Unit tests are changes after these changes in yast2, https://github.com/yast/yast-yast2/pull/1113. This should be fixed with https://github.com/yast/yast-yast2/pull/1114.

**Comment by ancorgs** [comment]
I did re-run the cirrus unit tests after merging https://github.com/yast/yast-yast2/pull/1114 and they are green now. I can't see a button to re-run the Travis tests.

**Review by ancorgs — APPROVED** [review]
LGTM now

**Pull request merged**